File: sol/2434/2434.py
class Solution:
    def robotWithString(self, s: str) -> str:
        n = len(s)
        smallestIdxFrom = [n - 1] * n
        for i in range(n - 2, -1, -1):
            smallestIdxFrom[i] = smallestIdxFrom[i + 1]
            if s[i] <= s[smallestIdxFrom[i]]:
                smallestIdxFrom[i] = i
        t = []
        res = ''
        ptr = 0
        while ptr < n:
            while smallestIdxFrom[ptr] != ptr:
                t.append(s[ptr])
                ptr += 1
            res += s[ptr]
            ptr += 1
            if ptr == n:
                break
            c = s[smallestIdxFrom[ptr]]
            while len(t) != 0 and t[-1] <= c:
                res += t.pop()
        res += ''.join(t[::-1])
        return res

File: sol/2434/2434.cpp
struct Solution {
    string robotWithString(string_view s) {
        int n = s.size();
        vector<int> smallest(n, n - 1);
        for (int i = n - 2; i >= 0; --i){
            smallest[i] = smallest[i + 1];
            if (s[i] <= s[smallest[i]]){
                smallest[i] = i;
            }
        }
        string t, res;
        int ptr = 0;
        while (ptr < n){
            while (smallest[ptr] != ptr){
                t.push_back(s[ptr++]);
            }
            res.push_back(s[ptr++]);
            if (ptr == n){
                break;
            }
            char c = s[smallest[ptr]];
            while (!t.empty() && t.back() <= c){
                res.push_back(t.back());
                t.pop_back();
            }
        }
        res.insert(res.end(), t.rbegin(), t.rend());
        return res;
    }
};
